Fix init of delta safe height (for G29, G33, etc.)

Bob-the-Kuhn opened the pull request:
The first probe point, when doing **G29** with **AUTO_BED_LEVELING_BILINEAR** enabled, moves the head to the extreme XY position via the following steps:

1. Lower Z to **Z_MAX_POS**
2. Lower Z to **delta_clip_start_height**
3. Move to XY  location of the point to be probed
4. Lower Z until Z probe fires

If **Z_MAX_POS** is too high then the XY move of the first G29 bilinear probe may hit the physical endstop which causes lost steps which throws off the resulting leveling grid.

On my Tevo Little Monster delta printer I have  **Z_MAX_POS** set to the height reported by G33.  When I do the G29 command, I see my gamma carriage move lower and the other two move higher, eventually hitting the physical limit which causes audible noise and lost steps.  After that the XYZ positions are invalid and the resulting leveling matrix is wrong.

I'm assuming **Z_MAX_POS** being set to the height reported by G33 is the norm.

My proposed fix is to add an offset to **delta_clip_start_height**.

**FROM:**
`float delta_clip_start_height = Z_MAX_POS`

**TO:**
```
#if !ENABLED(DELTA_FULL_XY_HEIGHT_OFFSET)
  #define DELTA_FULL_XY_HEIGHT_OFFSET 20;
#endif
float delta_clip_start_height = Z_MAX_POS - DELTA_FULL_XY_HEIGHT_OFFSET;  //G33 height is sometimes a little too high
```

---

**TESTING**

I've tested this on my Tevo Little Monster delta printer using bilinear and UBL leveling.

---

**OPTION**

Put the offset into the example delta configiurations rather than into code.  For  this I'd envision something like:

**FROM:**
```
#define Z_MAX_POS MANUAL_Z_HOME_POS
#define MANUAL_Z_HOME_POS DELTA_HEIGHT // Distance between the nozzle to printbed after homing
#define DELTA_HEIGHT 485.75               // (mm) Get this value from G33 auto calibrate
```

**TO**:
```
#define Z_MAX_POS MANUAL_Z_HOME_POS - DELTA_FULL_XY_HEIGHT_OFFSET
#define DELTA_FULL_XY_HEIGHT_OFFSET 2  // always 0 or positive - Distance from home position to where
                                       // full XY motion can start.  On some delta printers the head
                                       // can not extend to the full X/Y reach when at the home position
                                       // because that would force a carriage into the physical Z stop.
#define MANUAL_Z_HOME_POS DELTA_HEIGHT // Distance between the nozzle to printbed after homing
#define DELTA_HEIGHT 485.75               // (mm) Get this value from G33 auto calibrate
```

---

**OPTION**

Another way of attacking this is to change the order of moves in the G29 first probe.  Continue the Z movement down to the probe deploy point (some safe probe height) and then do the XY movement and then do the probing..  This is a common routine so I'm not sure if the new sequence is acceptable in all leveling schemes.

**FROM**
1. Lower Z to **Z_MAX_POS**
2. Lower Z to **delta_clip_start_height**
3. Move to XY  location of the point to be probed
4. Lower Z until Z probe fires

**TO**
1. Lower Z to probe deploy point (some safe probe height)
2. Move to XY  location of the point to be probed
3. Lower Z until Z probe fires

Bob-the-Kuhn committed:
create offset in case G33 height is too high

If Z_MAX_POS (which is usually set to the G33 height) is too high then the first G29 bilinear probe point may hit the physical endstop which causes lost steps which throws off the resulting leveling grid.

thinkyhead committed:
fix syntax

thinkyhead commented:
Please add serial debugging in `motion.cpp` after the line…
```cpp
delta_clip_start_height = soft_endstop.max[axis] - delta_safe_distance_from_top();
```
…because this is supposed to be executed during `setup()` after EEPROM settings have been loaded (or reset). If it is not getting called at boot time then it should be considered a bug.

Bob-the-Kuhn commented:
Does the **DEBUGGING(LEVELING)** statement 28 lines later already do what you want?

Bob-the-Kuhn committed:
use sytax suggested by Thinkyhead

Bob-the-Kuhn committed:
update local branch from PR

Bob-the-Kuhn committed:
different solution - move init & replicate logic

1 - eliminate old fix
2 - duplicate update_software_endstops logic in endstops.init() when no soft endstops

Bob-the-Kuhn commented:
Yes, it is a bug.  If no soft endstops are enabled then `delta_clip_start_height` is never initialized to a safe value by the `update_software_endstops()` routine.

My proposed solution is to replicate the `update_software_endstops()` logic for `delta_clip_start_height` at the end of the `Endstops::init`  routine.

I've tested this on my delta printer.  The G29 motion is the same whether soft endstops are enabled or not.

thisiskeithb commented:
Related issue: https://github.com/MarlinFirmware/Marlin/issues/19166

thinkyhead committed:
use refresh_delta_clip_start_height
